common/src/main/java/com/microsoft/identity/common/internal/controllers/BrokerOperationExecutor.java

```diff
@@ -0,0 +1,203 @@
+import com.microsoft.identity.common.exception.ErrorStrings;
+import com.microsoft.identity.common.internal.broker.ipc.IIpcStrategy;
+import com.microsoft.identity.common.internal.broker.ipc.BrokerOperationBundle;
+import com.microsoft.identity.common.internal.commands.parameters.CommandParameters;
+import com.microsoft.identity.common.internal.logging.Logger;
+import com.microsoft.identity.common.internal.telemetry.Telemetry;
+import com.microsoft.identity.common.internal.telemetry.events.ApiEndEvent;
+import com.microsoft.identity.common.internal.telemetry.events.ApiStartEvent;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Classes for executing IPC service operations.
+ * It takes in a list of IIpcStrategy, and will try to connect to the broker with each strategy. one by one.
+ * - If the current strategy succeeds, it will return the result right away.
+ * - If the service returns an error, it will return the result right away.
+ * - If the current strategy fails to connect to the targeted service, it will try the next one until the list is exhausted.
+ */
+public class BrokerOperationExecutor {
+
+    private static final String TAG = BrokerOperationExecutor.class.getSimpleName();
+
+    /**
+     * Info of a service operation to be performed with available strategies.
+     */
+    public interface BrokerOperation<T> {
+
+        /**
+         * Trigger any prerequisite works before making the actual request.
+         * This was added because the existing MSAL-Broker logic has a separate hello() call.
+         */
+        void performPrerequisites(@NonNull IIpcStrategy strategy) throws BaseException;
+
+        /**
+         * Gets a BrokerOperationBundle bundle to pass to each IpcStrategies.
+         */
+        @NonNull BrokerOperationBundle getBundle();
+
+        /**
+         * Extracts the result object from a bundle returned by an IpcStrategy.
+         * If the broker returns an error, this will throw an exception.
+         */
+        @NonNull T extractResultBundle(@Nullable final Bundle resultBundle) throws BaseException;
+
+        /**
+         * Returns method name (for logging/telemetry purpose).
+         */
+        @NonNull String getMethodName();
+
+        /**
+         * ID of the telemetry API event associated to this strategy task.
+         * If this value returns null, no telemetry event will be emitted.
+         */
+        @Nullable String getTelemetryApiId();
+
+        /**
+         * A method that will be invoked before the success event is emitted.
+         * If the calling operation wants to put any value in the success event, put it here.
+         */
+        void putValueInSuccessEvent(ApiEndEvent event, T result);
+    }
+
+    private final List<IIpcStrategy> mStrategies;
+
+    /**
+     * @param strategies list of IIpcStrategy to be invoked.
+     */
+    public BrokerOperationExecutor(@NonNull final List<IIpcStrategy> strategies) {
```

Comment:
Thought experiment: should this be specified in a configuration?

Reply:
IMO, this kind of stuff is too low-level and we should decide for them - with telemetry.
(e.g. once bind service is 'on the way out' in MSAL, there's no reason to let the devs pick it. we should just remove it).